clippy_lints/src/empty_with_brackets.rs

```diff
@@ -97,22 +115,94 @@ impl EarlyLintPass for EmptyWithBrackets {
         }
     }
 
-    fn check_variant(&mut self, cx: &EarlyContext<'_>, variant: &Variant) {
+    fn check_variant(&mut self, cx: &LateContext<'_>, variant: &Variant<'_>) {
+        // Don't lint pub enums
+        if cx.effective_visibilities.is_reachable(variant.def_id) {
+            return;
+        }
+
         let span_after_ident = variant.span.with_lo(variant.ident.span.hi());
 
-        if has_brackets(&variant.data) && has_no_fields(cx, &variant.data, span_after_ident) {
-            span_lint_and_then(
+        if has_no_fields(cx, &variant.data, span_after_ident) {
+            match variant.data {
+                VariantData::Struct { .. } => {
+                    span_lint_and_then(
+                        cx,
+                        EMPTY_ENUM_VARIANTS_WITH_BRACKETS,
+                        span_after_ident,
+                        "enum variant has empty brackets",
+                        |diagnostic| {
+                            diagnostic.span_suggestion_hidden(
+                                span_after_ident,
+                                "remove the brackets",
+                                "",
+                                Applicability::MaybeIncorrect,
+                            );
+                        },
+                    );
+                },
+                VariantData::Tuple(.., local_def_id) => {
+                    self.empty_tuple_enum_variants
+                        .entry(local_def_id)
+                        .or_insert(Usage::Unused {
+                            redundant_use_sites: vec![],
+                        });
+                },
+                VariantData::Unit(..) => {},
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(def_id) = check_expr_for_enum_as_function(expr) {
+            if let Some(parentheses_span) = call_parentheses_span(cx.tcx, expr)
+                && let Some(Usage::Unused {
+                    ref mut redundant_use_sites,
+                }) = self.empty_tuple_enum_variants.get_mut(&def_id)
```

Comment:
Could also insert `Usage::Unused { redundant_use_sites: vec![parentheses_span] }` in the `None` case similar to the entry above so that we still warn when the use expression is visited before its definition:
```rs
X::V();
enum X {
  V()
}
```